wfe/web-front-end.go

```diff
@@ -838,6 +863,86 @@ func (wfe *WebFrontEndImpl) NewCertificate(response http.ResponseWriter, request
 	}
 }
 
+// NewCertificateRequest is the asynchronous interface used by clients to
+// request the issuance of a cert for an authorized identifier.
+func (wfe *WebFrontEndImpl) NewCertificateRequest(response http.ResponseWriter, request *http.Request) {
+	logEvent := wfe.populateRequestEvent(request)
+	defer wfe.logRequestDetails(&logEvent)
+
+	body, _, reg, err := wfe.verifyPOST(request, true, core.ResourceNewCert)
+	if err != nil {
+		logEvent.Error = err.Error()
+		respMsg := malformedJWS
+		respCode := statusCodeFromError(err)
+		if _, ok := err.(core.NoSuchRegistrationError); ok {
+			respMsg = unknownKey
+			respCode = http.StatusForbidden
+		}
+		wfe.sendError(response, respMsg, err, respCode)
+		return
+	}
+	logEvent.Requester = reg.ID
+	logEvent.Contacts = reg.Contact
+	// Any version of the agreement is acceptable here. Version match is enforced in
+	// wfe.Registration when agreeing the first time. Agreement updates happen
+	// by mailing subscribers and don't require a registration update.
+	if reg.Agreement == "" {
+		logEvent.Error = "Must agree to subscriber agreement before any further actions"
+		wfe.sendError(response, logEvent.Error, nil, http.StatusForbidden)
+		return
+	}
+
+	var certificateRequest core.AcmeCertificateRequest
+	if err = json.Unmarshal(body, &certificateRequest); err != nil {
+		logEvent.Error = err.Error()
+		wfe.sendError(response, "Error unmarshaling certificate request", err, http.StatusBadRequest)
+		return
+	}
+	wfe.logCsr(request, certificateRequest, reg)
+	// Check that the key in the CSR is good. This will also be checked in the CA
+	// component, but we want to discard CSRs with bad keys as early as possible
+	// because (a) it's an easy check and we can save unnecessary requests and
+	// bytes on the wire, and (b) the CA logs all rejections as audit events, but
+	// a bad key from the client is just a malformed request and doesn't need to
+	// be audited.
+	if err = core.GoodKey(certificateRequest.CSR.PublicKey); err != nil {
+		logEvent.Error = err.Error()
+		wfe.sendError(response, "Invalid key in certificate request", err, http.StatusBadRequest)
+		return
+	}
+	logEvent.Extra["CSRDNSNames"] = certificateRequest.CSR.DNSNames
+	logEvent.Extra["CSREmailAddresses"] = certificateRequest.CSR.EmailAddresses
+	logEvent.Extra["CSRIPAddresses"] = certificateRequest.CSR.IPAddresses
+
+	// Create new certificate and return
+	// TODO IMPORTANT: The RA trusts the WFE to provide the correct key. If the
```

Comment:
I know this text predates the rules, but since we're copying it here, we should open an issue and mark it `TODO(xxx)`